File: 02_Object_Oriented_Programming/02_Class_Relationships/03_Aggregation/python/team.py
"""
Aggregation — Team class in Python.

A Team aggregates Players — holds references but does NOT own lifecycle.
"""

from __future__ import annotations
from player import Player


class Team:
    """A team that aggregates externally-created players."""

    def __init__(self, team_name: str) -> None:
        self._team_name = team_name
        self._players: list[Player] = []

    @property
    def team_name(self) -> str:
        return self._team_name

    @property
    def players(self) -> list[Player]:
        return list(self._players)  # return copy

    def add_player(self, player: Player) -> None:
        self._players.append(player)

    def remove_player(self, player: Player) -> None:
        self._players.remove(player)

    def show_team(self) -> None:
        print(f"Team {self._team_name} has players:")
        for p in self._players:
            print(f"  - {p}")

    def __str__(self) -> str:
        return f"Team({self._team_name}, size={len(self._players)})"

File: 02_Object_Oriented_Programming/02_Class_Relationships/03_Aggregation/javascript/team.js
/**
 * Aggregation — Team class in JavaScript.
 *
 * A Team aggregates Players — holds references but does NOT own lifecycle.
 */

class Team {
  #teamName;
  #players;

  constructor(teamName) {
    this.#teamName = teamName;
    this.#players = [];
  }

  get teamName() {
    return this.#teamName;
  }

  get players() {
    return [...this.#players]; // return copy
  }

  addPlayer(player) {
    this.#players.push(player);
  }

  removePlayer(player) {
    const idx = this.#players.indexOf(player);
    if (idx !== -1) this.#players.splice(idx, 1);
  }

  showTeam() {
    console.log(`Team ${this.#teamName} has players:`);
    for (const p of this.#players) {
      console.log(`  - ${p}`);
    }
  }

  toString() {
    return `Team(${this.#teamName}, size=${this.#players.length})`;
  }
}

module.exports = Team;
